opentasks-provider/src/main/java/org/dmfs/provider/tasks/processors/instances/Validating.java

```diff
@@ -43,50 +69,79 @@ public Validating(EntityProcessor<InstanceAdapter> delegate)
     public InstanceAdapter insert(SQLiteDatabase db, InstanceAdapter entityAdapter, boolean isSyncAdapter)
     {
         validate(entityAdapter, isSyncAdapter);
-        return mDelegate.insert(db, entityAdapter, isSyncAdapter);
+
+        Optional<Long> instanceId = new NullSafe<>(entityAdapter.taskAdapter().valueOf(TaskAdapter.ORIGINAL_INSTANCE_ID));
+        if (instanceId.isPresent())
+        {
+            String timeStampString = Long.toString(entityAdapter.taskAdapter().valueOf(TaskAdapter.ORIGINAL_INSTANCE_TIME).getTimestamp());
+            // there better be no instance for this yet, otherwise this is going to fail
+            try (Cursor c = db.query(
+                    TaskDatabaseHelper.Tables.INSTANCE_VIEW,
+                    new String[] { TaskContract.Instances._ID },
+                    // find any instance which refers to the given original ID and has the same instance time
+                    // for recurring tasks this matches the INSTANCE_ORIGINAL_TIME, for non-recurring tasks this matches start or due (whichever is present).
+                    String.format("(%1$s == ? or %2$s == ?) and (%3$s == ? or %3$s is null and %4$s == ? or %3$s is null and %4$s is null and %5$s == ?) ",
```

Comment:
This query is a bit difficult to follow for me, would you consider splitting it to two queries (recurring, non-recurring)?
Maybe not effectively run two queries, just split the sql query string into two - if that can be done - like this:
```java
// check for existing non-recurring task:
String.format("........", TaskContract.Instances.TASK_ID, TaskContract.Instances.INSTANCE_START, TaskContract.Instances.INSTANCE_DUE)
+ " or " + 
// check for existing recurring instance:
String.format("....", TaskContract.Instances.ORIGINAL_INSTANCE_ID, TaskContract.Instances.INSTANCE_ORIGINAL_TIME,
```

By the way, for me it's not clear whether `String.format()` or string concatenation is easier to read for this sql statements. (Kotlin's string template would be of course be better.)

Reply:
neither way appears to be more readable to me, let's keep it like that until we can replace it with something more sophisticated (i.e. Anko or "SQLitePal" ;-) ). As a first step we could port the ContentPal predicates. That shouldn't be to difficult but can help a lot to write meaningful select statements.

Reply:
There is only one test for the non-recurring case currently, right? Can we cover the other case(s) as well, or add TODOs in the test?

Reply:
I can see that select string builder with the predicates would be useful here.
I've just checked and Anko SQLite doesn't seem to have anything for that.

Reply:
The recurrence tests are part of another story (the one which implements the recurrence expansion). I already implemented most of them.